src/Voron/Util/PtrSize.cs

```diff
@@ -1,8 +1,100 @@
-﻿namespace Voron.Util
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Sparrow;
+
+namespace Voron.Util
 {
+    [StructLayout(LayoutKind.Explicit)]
     public unsafe struct PtrSize
     {
-        public byte* Ptr;
-        public int Size;
+        private const uint ValueMask = 0x80000000;
+
+        [FieldOffset(0)]
+        public readonly byte* Ptr;
+
+        [FieldOffset(0)]
+        public readonly long Value;
+
+        [FieldOffset(8)]
+        private readonly uint InternalSize;
+            
+        
+        public int Size
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get { return (int)(InternalSize & ~ValueMask); }
+        }
+
+        public bool IsValue
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get { return (InternalSize & ValueMask) != 0; }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private PtrSize(long value, uint size)
+        {
+            Debug.Assert(size <= 8);
+
+            Ptr = null;
+            Value = value;
+            InternalSize = ValueMask | size;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private PtrSize(byte* ptr, uint size)
+        {
+            Debug.Assert(size <= int.MaxValue >> 1);
+
+            Value = 0;
+            Ptr = ptr;
+            InternalSize = size;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static PtrSize Create<T>(T value) where T : struct 
+        {
+            if (typeof(T) == typeof(ulong) || typeof(T) == typeof(long))
+            {
+                return new PtrSize((long)(object)value, sizeof(long));
```

Comment:
Are those object cast being erased by the JIT? If not we do boxing & unboxing here.